### PR TITLE
Add extension for changed lines in GitLab MR

### PR DESCRIPTION
**Pull request by davidbilik**
Since GitLab does not provide number of added/deleted lines
in Merge requests add artificial properties that are computed
through `git` command.

Related: #GH-101

### CODE REVIEW

File: danger-kotlin-library/src/main/kotlin/systems/danger/kotlin/GitLabKtx.kt [hunk 1]
```diff
@@ -0,0 +1,47 @@
+package systems.danger.kotlin
+
+// extensions on entities from [GitLab] file
+
+/**
+ * Changed lines in this MR
+ */
+val GitLabMergeRequest.changedLines: MergeRequestChangedLines
+    get() {
+        val commandRawOutput = Utils().exec("git diff --numstat ${diffRefs.startSha} ${diffRefs.headSha}")
```
Comment (davidbilik) [inline danger-kotlin-library/src/main/kotlin/systems/danger/kotlin/GitLabKtx.kt:10]: I'm not sure what is the best way how to use `Utils` sine it's a class and not an `object`. Creating new instance here seems wrong.

File: danger-kotlin-library/src/main/kotlin/systems/danger/kotlin/GitLabKtx.kt [hunk 2]
```diff
@@ -0,0 +1,47 @@
+package systems.danger.kotlin
```
Comment (davidbilik) [inline danger-kotlin-library/src/main/kotlin/systems/danger/kotlin/GitLabKtx.kt:1]: Also I'm not sure what is the best place to put this. I see `GitLab` object only as a data holder and as a direct mapping of it's api representation so I've placed it to separate file and use it as an extension but I don't know if it follows this project's code style 🤷

**Comment by danger-public** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  Any Changes To Library Code Should Be Reflected In The Changelog.

Please Consider Adding A Note There And Adhere To The [Changelog Guidelines](https://github.com/Moya/contributors/blob/master/Changelog%20Guidelines.md).
  </td>
    </tr>
  </tbody>
</table>



<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems">Danger Kotlin</a> against dfd65b2dd66a8d2e298da5d282a08cd5e98c9dbe
</p>

File: danger-kotlin-library/src/main/kotlin/systems/danger/kotlin/GitLabKtx.kt (hunk 2 above)
Comment (f-meloni) [reply danger-kotlin-library/src/main/kotlin/systems/danger/kotlin/GitLabKtx.kt:1]: I'm fine either way, I would personally have used the GitLab file, but no strong opinion, don't know if @gianluz has a strong opinion.

What I'm thinking thought is that this is generic enough to live in `Git` (given it uses git to retrieve the count).
In this way it could be used also from Bitbucket

File: danger-kotlin-library/src/main/kotlin/systems/danger/kotlin/GitLabKtx.kt (hunk 1 above)
Comment (f-meloni) [reply danger-kotlin-library/src/main/kotlin/systems/danger/kotlin/GitLabKtx.kt:10]: I agree, it feels wrong, the correct way to access utils should be from the DangerDSL, but to access that you need to be in the Dangerfile.
Then what I would do is create a simple class that can run commands (example from danger swift https://github.com/danger/swift/blob/master/Sources/DangerShellExecutor/ShellExecutor.swift), and use that both in utils and here.

Maybe this could also help us to find a way to test this?

File: danger-kotlin-library/src/main/kotlin/systems/danger/kotlin/GitLabKtx.kt (hunk 2 above)
Comment (davidbilik) [reply danger-kotlin-library/src/main/kotlin/systems/danger/kotlin/GitLabKtx.kt:1]: I think that `Git` would be ideal place as well but I've used `GitLab` because it provides me metadata about the commits which I need to pass to a `git` command. I have not found any way how to access this info from other place. Do you have any suggestion?

Comment (davidbilik) [reply danger-kotlin-library/src/main/kotlin/systems/danger/kotlin/GitLabKtx.kt:1]: Ok, I might have a hint. I can see that in the JSON from danger-js is also list of commits, that is not parsed currently. I can run `git diff --numstats` for this commits and aggregate changed lines based on that (I think that is the similar approach that danger-js has). I will investigate more after later today :)

Comment (gianluz) [reply danger-kotlin-library/src/main/kotlin/systems/danger/kotlin/GitLabKtx.kt:1]: i'm happy to use separate files for the extensions, we just need to keep conventions once started.. eg: `GitHubKtx.kt` `GitKtx.kt` ...

File: danger-kotlin-library/src/main/kotlin/systems/danger/kotlin/GitLabKtx.kt (hunk 1 above)
Comment (gianluz) [reply danger-kotlin-library/src/main/kotlin/systems/danger/kotlin/GitLabKtx.kt:10]: agree as well, maybe convert Utils to an `object` ?

Comment (davidbilik) [reply danger-kotlin-library/src/main/kotlin/systems/danger/kotlin/GitLabKtx.kt]: I've left this unprocessed for now and I would focus on that in the next MR

File: danger-kotlin-library/src/main/kotlin/systems/danger/kotlin/GitLabKtx.kt (hunk 2 above)
Comment (davidbilik) [reply danger-kotlin-library/src/main/kotlin/systems/danger/kotlin/GitLabKtx.kt]: Ok guys, take a look now (I've force pushed for cleaner git history). I've basically took the provided list of commits from the JSON from danger-js and I've computed this statistics for each of the commit and then just sum it up. It works when I try to test it on my GitLab MRs. I've also made it as extensions over `Git` which is more generic

File: danger-kotlin-library/src/main/kotlin/systems/danger/kotlin/GitLabKtx.kt (hunk 1 above)
Comment (f-meloni) [reply danger-kotlin-library/src/main/kotlin/systems/danger/kotlin/GitLabKtx.kt]: We could make (I’m fine to see it next though) a Cmd class like the one on the runner that we could share wherever we have to execute shell code

File: danger-kotlin-library/src/main/kotlin/systems/danger/kotlin/GitKtx.kt [hunk 3]
```diff
@@ -0,0 +1,59 @@
+package systems.danger.kotlin
+
+// extensions over [Git] object
+
+/**
+ * Changed lines in this MR
```
Comment (f-meloni) [inline danger-kotlin-library/src/main/kotlin/systems/danger/kotlin/GitKtx.kt:6]: Nitpick: I think PR is the more used term.

File: danger-kotlin-library/src/main/kotlin/systems/danger/kotlin/GitKtx.kt [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+ */
+val Git.changedLines: PullRequestChangedLines
+    get() {
+        // TODO replace with more generic way to process commands
+        val utils = Utils()
+        val additionDeletionPairs = commits.mapNotNull { it.sha }
+            .map { sha ->
+                computeChangedLinesForCommit(utils, sha)
```
Comment (f-meloni) [inline danger-kotlin-library/src/main/kotlin/systems/danger/kotlin/GitKtx.kt:14]: Given the list is ordered, wouldn’t make sense to pass directly first and last commit and get the diff between them? That would make it a lot more efficient

File: danger-kotlin-library/src/main/kotlin/systems/danger/kotlin/GitKtx.kt (hunk 3 above)
Comment (davidbilik) [reply danger-kotlin-library/src/main/kotlin/systems/danger/kotlin/GitKtx.kt]: Yea you're right, I even use it in the wrapper class c22ea384100a297780be7b0499ba6687a407e0b6

File: danger-kotlin-library/src/main/kotlin/systems/danger/kotlin/GitKtx.kt (hunk 4 above)
Comment (davidbilik) [reply danger-kotlin-library/src/main/kotlin/systems/danger/kotlin/GitKtx.kt]: Actually you're right. We get only list of commits that are included in this PR but I can get first and last^1 and it works as well. Rewritten in c22ea384100a297780be7b0499ba6687a407e0b6

File: danger-kotlin-library/src/main/kotlin/systems/danger/kotlin/GitLabKtx.kt (hunk 1 above)
Comment (davidbilik) [reply danger-kotlin-library/src/main/kotlin/systems/danger/kotlin/GitLabKtx.kt]: Ok, I've prepared something, I will create a PR once this one is resolved :) [Here](https://github.com/davidbilik/kotlin/commit/0dca794954fab43b58d74eb6c513dfb50d6c46f2) is a commit in my fork

File: danger-kotlin-library/src/main/kotlin/systems/danger/kotlin/GitKtx.kt [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+ * Reference to a SHA of head commit of this PR
+ */
+val Git.headSha: String?
+    get() = commits.first().sha
+
+/**
+ * Reference to a SHA of base commit of this PR
+ */
+val Git.baseSha: String?
+    get() = "${commits.last().sha}^1"
```
Comment (f-meloni) [inline danger-kotlin-library/src/main/kotlin/systems/danger/kotlin/GitKtx.kt:52]: Why `^1`? I think your base is the actual `commits.last().sha`
If you use this command here you would skip the changes from `c22ea38` I think, am I missing something?

Comment (davidbilik) [reply danger-kotlin-library/src/main/kotlin/systems/danger/kotlin/GitKtx.kt:52]: Maybe I am missing something, but if I take this PR as an example, so here are three commits:
* 7370a5fa2c0720c3e38322b75603687aea0f1d6a
* 3e1bd1539b12d885f3943995bbfcc9c275bf11a2
* c22ea384100a297780be7b0499ba6687a407e0b6

if I compute diff like
```
git diff --numstat c22ea384100a297780be7b0499ba6687a407e0b6 7370a5fa2c0720c3e38322b75603687aea0f1d6a
```
the result is 
```
0       2       CHANGELOG.md
19      20      danger-kotlin-library/src/main/kotlin/systems/danger/kotlin/GitKtx.kt
```
it will return changes that happened from the commit c22ea384100a297780be7b0499ba6687a407e0b6 to 7370a5fa2c0720c3e38322b75603687aea0f1d6a but it does not take changes in the 7370a5fa2c0720c3e38322b75603687aea0f1d6a into account. So if I take the changes from the parent of the first commit up until the last commit we know, in our case
```
git diff --numstat c22ea384100a297780be7b0499ba6687a407e0b6 7370a5fa2c0720c3e38322b75603687aea0f1d6a^1
```
it will return correctly
```
0       2       CHANGELOG.md
1       2       danger-kotlin-library/src/main/kotlin/systems/danger/kotlin/Git.kt
0       60      danger-kotlin-library/src/main/kotlin/systems/danger/kotlin/GitKtx.kt
3       3       danger-kotlin-library/src/main/kotlin/systems/danger/kotlin/Utils.kt
```
which if summed up matches the number [here](https://github.com/danger/kotlin/pull/102/commits)

Also if you would have only a single commit in the PR, eg. in our repo [here](https://github.com/AckeeCZ/ackee-gradle-plugin/pull/13/commits) the command
```
git diff --numstat 49ec200b25a1130cfb86933f41bbf79e40b997bf
```
produces
```
4       0       README.md
33      8       build-gradle-plugin/src/main/kotlin/cz/ackee/gradle/AckeePluginKotlin.kt
1       1       lib.properties
```
which does not match. But if I do this
```
git diff --numstat 49ec200b25a1130cfb86933f41bbf79e40b997bf 49ec200b25a1130cfb86933f41bbf79e40b997bf^1
```
it produces that
```
0       31      .github/workflows/bintray_upload.yml
4       12      gradle/bintray/bintray.gradle
```
which is correct. 

So maybe I am missing something or I am doing it ineffciently, but from my testing this method works. 

Basically I want to see all changes from the commit my branch is originating from up until the last commit in my PR. That's how I understand it

Comment (f-meloni) [reply danger-kotlin-library/src/main/kotlin/systems/danger/kotlin/GitKtx.kt:52]: Ok it makes sense, the last commit is then the actual first commit you made.
Now it makes sense, thank you :)
